### PR TITLE
Upgrade base64 crate

### PR DESCRIPTION
**Pull request by vojd**
### Checklist

* [x] I have read the [Contributor Guide](../../CONTRIBUTING.md)
* [x] I have read and agree to the [Code of Conduct](../../CODE_OF_CONDUCT.md)
* [x] I have added a description of my changes and why I'd like them included in the section below

### Description of Changes

Upgrade base64 0.13 to 0.21
Use `URL_SAFE_NO_PAD` for base64 encoding
Upgrade tokio (used by the examples) from 1.21 to 1.28
Add tests for `Provider` and new de-serializer for optional fields

Use https://github.com/EmbarkStudios/jsonwebtoken/tree/fix/pem-2.0 while waiting for https://github.com/Keats/jsonwebtoken/pull/303 to get merged. We need the updated pem library in order to use `base64 0.21`

### CODE REVIEW

**Review by MarcusGrass — APPROVED** [review]
Ship :shipit:!

**Comment by vojd** [comment]
To prevent using conflicting versions of `base64` I added a fork to `jsonwebtoken` which simply bumps `pem` from `1` to `2`.
Though we should wait for https://github.com/Keats/jsonwebtoken/pull/303 to get in and I feel a little reluctant to release a new version of `tame-oidc` before that PR gets in.

File: Cargo.toml [hunk 1]
```diff
@@ -21,18 +21,20 @@ path = "examples/embark.rs"
 [dependencies]
 base64 = "0.21"
 http = "0.2"
-jsonwebtoken = "8.1"
+# While waiting for https://github.com/Keats/jsonwebtoken/pull/303 to get in we need
+# to rely on this forked commit for the `pem` crate to use `base64 0.21`
+jsonwebtoken = { git = "https://github.com/EmbarkStudios/jsonwebtoken", rev = "ccd4394b49c1b92590ba6cca0f5d1ec44c813acd" }
```
Comment (repi) [inline Cargo.toml:26]: not possible to have git dependencies in crates published on crates.io. have to get this released before so one can reference a published version of the dependency here

**Review by repi — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 1 above)
Comment (vojd) [reply Cargo.toml:26]: > Though we should wait for https://github.com/Keats/jsonwebtoken/pull/303 to get in
> and I feel a little reluctant to release a new version of tame-oidc before that PR gets in.

That makes the decision even easier.

**Comment by vojd** [comment]
Will split up this PR into new ones instead.

**Pull request closed without merging**